### PR TITLE
[WIP] Custom OEmbed providers

### PR DESCRIPTION
**Pull request by allouis**
no-issue

This PR is open to share code with Kevin and receive feedback

### CODE REVIEW

File: core/server/services/oembed.js [hunk 1]
```diff
@@ -53,7 +59,26 @@ class OEmbed {
      */
     constructor({config, externalRequest}) {
         this.config = config;
-        this.externalRequest = externalRequest;
+        /** @type {IExternalRequest} */
+        this.externalRequest = async (url, requestConfig) => {
```
Comment (allouis) [inline core/server/services/oembed.js:63]: This wraps the external request with checks before & after for ip/localhost - this is so custom providers don't need to know/care about that. 

I want to check this looks correct, and if it does, then we should be able to remove these checks from the methods below and just use `this.externalRequest`

File: core/server/services/oembed.js [hunk 2]
```diff
@@ -274,6 +299,16 @@ class OEmbed {
         let data;
 
         try {
+            const urlObject = new URL(url);
+            for (const provider of this.customProviders) {
+                if (await provider.canSupportRequest(urlObject)) {
+                    const result = await provider.getOEmbedData(urlObject, this.externalRequest);
+                    if (result !== null) {
```
Comment (allouis) [inline core/server/services/oembed.js:306]: Allows fallthrough

File: core/server/services/nft-oembed.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    async getOEmbedData(url, externalRequest) {
+        const [match, transaction, asset] = url.pathname.match(OPENSEA_PATH_REGEX);
+        if (!match) {
+            return null;
+        }
+        const result = await externalRequest(`https://api.opensea.io/api/v1/asset/${transaction}/${asset}/`, {
+            json: true
+        });
+        return {
```
Comment (allouis) [inline core/server/services/nft-oembed.js:41]: I'm trying to stick to the OEmbed spec here - but I'm not sure if that's how you imagined it?

What's particularly tricky with the OEmbed spec is that we should provide the minimum width & height in pixels - which I've hardcoded to 1000 atm :grimacing:

File: core/server/services/nft-oembed.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * @implements ICustomProvider
+ */
+class NFTOEmbedProvider {
+    /**
+     * @param {object} dependencies
+     */
+    constructor(dependencies) {
+        this.dependencies = dependencies;
```
Comment (allouis) [inline core/server/services/nft-oembed.js:16]: placeholder

File: core/server/services/nft-oembed.js (hunk 3 above)
Comment (kevinansfield) [reply core/server/services/nft-oembed.js:41]: Yeah, we should stick the the oembed spec as much as possible. We don't actually use the width/height for anything at the moment other than to verify that we have what looks like a [valid oembed response here](https://github.com/TryGhost/Ghost/blob/main/core/server/services/oembed.js#L255-L257) so hardcoded value is fine for now 👍

File: core/server/services/oembed.js (hunk 1 above)
Comment (kevinansfield) [reply core/server/services/oembed.js:63]: 👍 LGTM and agree we can remove the other checks below

**Pull request closed without merging**